src/dataflow/src/source/kafka.rs

```diff
@@ -614,27 +681,114 @@ impl ControlPlaneInfo {
             },
         );
     }
+
+    /// Generates a timestamp that is guaranteed to be monotonically increasing.
+    /// This may require multiple calls to the underlying now() system method, which is not
+    /// guaranteed to increase monotonically
+    fn generate_next_timestamp(&mut self) -> u64 {
+        // TODO(ncrooks) - If someone does something silly like sets their
+        // system clock backward by an hour while mz is running,
+        // we will hang here for an hour.
+        let mut new_ts = 0;
+        while new_ts <= self.last_closed_ts {
+            let start = SystemTime::now();
+            new_ts = start
+                .duration_since(UNIX_EPOCH)
+                .expect("Time went backwards")
+                .as_millis() as u64;
+        }
+        assert!(new_ts > self.last_closed_ts);
+        self.last_closed_ts = new_ts;
+        new_ts
+    }
 }
 
 /// This function activates the necessary timestamping information when a source is first created
-/// These steps are only taken if a source actively reads data
-/// 1) it inserts an entry in the timestamp_history datastructure
-/// 2) it notifies the coordinator that timestamping should begin by inserting an entry in the
-/// timestamp_tx channel
-fn activate_source_timestamping<G>(config: &SourceConfig<G>, connector: KafkaSourceConnector) {
-    let prev = config
-        .timestamp_histories
-        .borrow_mut()
-        .insert(config.id.clone(), HashMap::new());
-    // Check that this is the first time this source id is registered
-    assert!(prev.is_none());
-    config.timestamp_tx.as_ref().borrow_mut().push((
-        config.id,
-        Some((
-            ExternalSourceConnector::Kafka(connector),
-            config.consistency.clone(),
-        )),
-    ));
+/// 1) for BYO sources, we notify the timestamping thread that a source has been created.
+/// 2) for a RT source, take no steps
+//
+fn activate_source_timestamping<G>(
+    config: &SourceConfig<G>,
+    connector: KafkaSourceConnector,
+) -> Option<TimestampChanges> {
+    if let Consistency::BringYourOwn(_) = config.consistency {
+        let prev = config
+            .timestamp_histories
+            .borrow_mut()
+            .insert(config.id.clone(), HashMap::new());
+        // Check that this is the first time this source id is registered
+        assert!(prev.is_none());
+        config.timestamp_tx.as_ref().borrow_mut().push((
+            config.id,
+            Some((
+                ExternalSourceConnector::Kafka(connector),
+                config.consistency.clone(),
+            )),
+        ));
+        return Some(config.timestamp_tx.clone());
+    }
+    None
+}
+
+/// This function spawns a new thread responsible for periodically polling Kafka metadata
```

Comment:
The comment isn't quite accurate -- this doesn't spawn a thread. It's just a plain function (which runs on a thread that somewhere else spawns)